[AutoFill Debugging] Refactor debug text extraction to work with non-Cocoa ports

whsieh opened the pull request:
#### b7f73eb8a257afa72ba6adec0ebb0bb19017eaad
<pre>
[AutoFill Debugging] Refactor debug text extraction to work with non-Cocoa ports
<a href="https://bugs.webkit.org/show_bug.cgi?id=300135">https://bugs.webkit.org/show_bug.cgi?id=300135</a>
<a href="https://rdar.apple.com/161919662">rdar://161919662</a>

Reviewed by Abrar Rahman Protyasha.

This patch moves logic to convert a `WebCore::TextExtraction::Item` into a hierarchical debug text
representation out of platform-specific Swift/ObjC code in `_WKTextExtraction.swift` and
`WKTextExtractionUtilities.mm`, and into a new `TextExtractionToStringConversion.cpp`, which is
platform agnostic.

This also allows us to skip the conversion from `WebCore::TextExtraction::Item` to a tree of
`WKTextExtractionItem` Objective-C objects for the purposes of extracting `-_debugText:`, and
instead directly convert from `WebCore::TextExtraction::Item` to a string representation.

Note that this new helper file is in `Source/WebKit/Shared`, so that it can be invoked from within
the web process if necessary, in a subsequent patch. For now, this helper is only invoked from the
UI process.

No change in behavior (beyond the lack of trailing newline in the output below); see below for more
details.

* LayoutTests/fast/text-extraction/debug-text-extraction-basic-expected.txt:

Rebaseline a layout test; the debug text extraction no longer has a trailing newline with the new
way of aggregating text results per line.

* Source/WebKit/Shared/Extensions/WebExtensionSQLiteStatement.h:
* Source/WebKit/Shared/TextExtractionToStringConversion.cpp: Added.
(WebKit::TextExtractionAggregator::TextExtractionAggregator):
(WebKit::TextExtractionAggregator::~TextExtractionAggregator):

Create a new RAII object to encapsulate the debug text conversion lifecycle. This is necessary
because `TextExtraction::Item` itself is just a struct (not ref-counted) that&apos;s passed back in the
completion handler of `WebPageProxy::requestTextExtraction`, but when recursively converting these
text extraction results, we may need to perform asynchronous filtering operations to verify that
(e.g.) text content in extracted text items represent user-visible text, and the
`TextExtractionFilter`&apos;s classifier considers the text valid. This was previously not necessary,
because we took advantage of the fact that we first converted `TextExtraction::Item` to ObjC objects
which we just retained while performing async filtering (and whose text we could mutate as needed).

This is created with the overall completion handler, which is invoked in the destructor. See
`convertToText` below for the usage.

(WebKit::TextExtractionAggregator::create):
(WebKit::TextExtractionAggregator::addResult):

Add a helper method to add results (give a line number, indentation level and comma-separated
components) to the final text extraction. Importantly, note that these results don&apos;t necessary come
in in order by line number, due to the fact that extraction for text nodes may involve additional
async validation.

(WebKit::TextExtractionAggregator::advanceToNextLine):

Add a helper method to advance to the next line (i.e. claiming the next line number) when dumping
text for a new item in the extraction result tree.

(WebKit::commaSeparatedString):
(WebKit::escapeString):
(WebKit::eventListenerTypesToStringArray):

Migrate several helper methods over from `_WKTextExtraction.swift`.

(WebKit::partsForItem):
(WebKit::addPartsForText):
(WebKit::addPartsForItem):
(WebKit::addTextRepresentationRecursive):

Implement the bulk of the algorithm here. This recursively traverses extraction results, adding
results to the `aggregator` in the process. This takes an async filtering callback (which returns a
`NativePromise`) which applies to all extracted text items. While any of these promises is pending,
the `aggregator` will be kept alive.

(WebKit::convertToText):
* Source/WebKit/Shared/TextExtractionToStringConversion.h: Copied from Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.h.
(WebKit::convertToText):
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _activeNativeMenuItemTitles]):

Make this return a list of item titles, now that we no longer use `WKTextExtractionPopupMenu`.

(-[WKWebView _debugTextWithConfiguration:completionHandler:]):

Reimplement this in terms of `-_requestTextExtractionInternal:completion:`; instead of getting a
`WKTextExtractionResult` and converting that into a text representation, we now go directly from
`TextExtraction::Item` to `NSString *`.

We also reimplement the async filtering logic here (i.e. passing text through
`WebKit::TextExtractionFilter` and `-_validateText:`) using the `TextExtractionFilterCallback` in
`convertToText`, which allows us to delete the now-unused helpers to recursively filter out text in
the ObjC extraction result tree, in `WKTextExtractionUtilities.mm`.

(-[WKWebView _requestTextExtractionInternal:completion:]):

Pull common logic to obtain an `TextExtraction::Item` out into a separate helper method.

(-[WKWebView _requestTextExtraction:completionHandler:]):
(-[WKWebView _validateText:inNode:completionHandler:]):
(-[WKWebView _popupMenuForTextExtractionResults]): Deleted.

Remove `WKTextExtractionPopupMenu` entirely, now that debug text extraction no longer depends on
producing a `WKTextExtractionResult` and asking it for a text representation.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.swift:
(WKTextExtractionEventListenerTypes.description): Deleted.
(eventListenerTypesAsArray(_:)): Deleted.
(String.escaped): Deleted.
(WKTextExtractionItem.textRepresentationRecursive(_:)): Deleted.
(WKTextExtractionItem.textRepresentationParts): Deleted.
(WKTextExtractionContainerItem.textRepresentationParts): Deleted.
(WKTextExtractionContentEditableItem.textRepresentationParts): Deleted.
(WKTextExtractionTextFormControlItem.textRepresentationParts): Deleted.
(WKTextExtractionLinkItem.textRepresentationParts): Deleted.
(WKTextExtractionTextItem.textRepresentationParts): Deleted.
(WKTextExtractionScrollableItem.textRepresentationParts): Deleted.
(WKTextExtractionSelectItem.textRepresentationParts): Deleted.
(WKTextExtractionImageItem.textRepresentationParts): Deleted.
(WKTextExtractionPopupMenu.textRepresentation): Deleted.
(WKTextExtractionResult.textRepresentation): Deleted.

Delete now-unused code (see `TextExtractionToStringConversion.cpp` above).

* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtractionInternal.h:
* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.h:
* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.mm:
(WebKit::filterTextRecursive): Deleted.
(WebKit::filterText): Deleted.

Moved into `WKWebView.mm` (see above).

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/301002@main">https://commits.webkit.org/301002@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6c1924f5ffbe33299bb386e0ba9bdce33ecd4380

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124593 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44265 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34999 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131432 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76536 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/293ff1eb-30cf-4a46-a2dd-d463fd0e4e85) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44969 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52836 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94784 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62853 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/01d167ea-3326-41b5-8eef-89caa41eeebf) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127547 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35851 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111422 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75355 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/5a43e053-5f90-4359-bd9d-bd32a3e42623) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34786 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29577 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74911 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105606 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29808 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134100 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51443 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39263 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103262 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51853 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107644 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103041 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26240 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48402 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26666 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48375 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51316 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57112 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50715 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54070 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52404 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/Shared/TextExtractionToStringConversion.h:
```diff
@@ -25,28 +25,28 @@
 
-#endif // ENABLE(LOGD_BLOCKING_IN_WEBCONTENT)
+using TextExtractionFilterPromise = NativePromise<String, void>;
+using TextExtractionFilterCallback = Function<Ref<TextExtractionFilterPromise>(const String&, std::optional<WebCore::NodeIdentifier>)>;
```
```suggestion
using TextExtractionFilterCallback = Function<Ref<TextExtractionFilterPromise>(const String&, std::optional<WebCore::NodeIdentifier>&&)>;
```

aprotyas commented on Source/WebKit/Shared/TextExtractionToStringConversion.cpp:
```diff
@@ -0,0 +1,411 @@
+};
+
+class TextExtractionAggregator : public RefCounted<TextExtractionAggregator> {
+    WTF_MAKE_NONCOPYABLE(TextExtractionAggregator);
```
nit: also TZone allocated?

aprotyas commented on Source/WebKit/Shared/TextExtractionToStringConversion.cpp:
```diff
@@ -0,0 +1,411 @@
+
+    ~TextExtractionAggregator()
+    {
+        if (auto completion = std::exchange(m_completion, { }))
```
You don't need to do the exchange, right? Just call the handler if it hasn't been already?
```suggestion
        if (m_completion)
            m_completion(...);
```

whsieh replied:
Ah, I generally do this to ensure there's no possibility of reentrancy (though in this case, that wouldn't happen anyways). I'll just remove the exchange.

aprotyas commented on Source/WebKit/Shared/TextExtractionToStringConversion.cpp:
```diff
@@ -0,0 +1,411 @@
+    auto attributeKeys = copyToVector(item.ariaAttributes.keys());
+    std::sort(attributeKeys.begin(), attributeKeys.end(), [](const String& a, const String& b) {
+        return codePointCompareLessThan(a, b);
+    });
```
Slightly less typing

```suggestion
    std::ranges::sort(attributeKeys, codePointCompareLessThan);
```

whsieh committed:
[AutoFill Debugging] Refactor debug text extraction to work with non-Cocoa ports
https://bugs.webkit.org/show_bug.cgi?id=300135
rdar://161919662

Reviewed by Abrar Rahman Protyasha.

This patch moves logic to convert a `WebCore::TextExtraction::Item` into a hierarchical debug text
representation out of platform-specific Swift/ObjC code in `_WKTextExtraction.swift` and
`WKTextExtractionUtilities.mm`, and into a new `TextExtractionToStringConversion.cpp`, which is
platform agnostic.

This also allows us to skip the conversion from `WebCore::TextExtraction::Item` to a tree of
`WKTextExtractionItem` Objective-C objects for the purposes of extracting `-_debugText:`, and
instead directly convert from `WebCore::TextExtraction::Item` to a string representation.

Note that this new helper file is in `Source/WebKit/Shared`, so that it can be invoked from within
the web process if necessary, in a subsequent patch. For now, this helper is only invoked from the
UI process.

No change in behavior (beyond the lack of trailing newline in the output below); see below for more
details.

* LayoutTests/fast/text-extraction/debug-text-extraction-basic-expected.txt:

Rebaseline a layout test; the debug text extraction no longer has a trailing newline with the new
way of aggregating text results per line.

* Source/WebKit/Shared/Extensions/WebExtensionSQLiteStatement.h:
* Source/WebKit/Shared/TextExtractionToStringConversion.cpp: Added.
(WebKit::TextExtractionAggregator::TextExtractionAggregator):
(WebKit::TextExtractionAggregator::~TextExtractionAggregator):

Create a new RAII object to encapsulate the debug text conversion lifecycle. This is necessary
because `TextExtraction::Item` itself is just a struct (not ref-counted) that's passed back in the
completion handler of `WebPageProxy::requestTextExtraction`, but when recursively converting these
text extraction results, we may need to perform asynchronous filtering operations to verify that
(e.g.) text content in extracted text items represent user-visible text, and the
`TextExtractionFilter`'s classifier considers the text valid. This was previously not necessary,
because we took advantage of the fact that we first converted `TextExtraction::Item` to ObjC objects
which we just retained while performing async filtering (and whose text we could mutate as needed).

This is created with the overall completion handler, which is invoked in the destructor. See
`convertToText` below for the usage.

(WebKit::TextExtractionAggregator::create):
(WebKit::TextExtractionAggregator::addResult):

Add a helper method to add results (give a line number, indentation level and comma-separated
components) to the final text extraction. Importantly, note that these results don't necessary come
in in order by line number, due to the fact that extraction for text nodes may involve additional
async validation.

(WebKit::TextExtractionAggregator::advanceToNextLine):

Add a helper method to advance to the next line (i.e. claiming the next line number) when dumping
text for a new item in the extraction result tree.

(WebKit::commaSeparatedString):
(WebKit::escapeString):
(WebKit::eventListenerTypesToStringArray):

Migrate several helper methods over from `_WKTextExtraction.swift`.

(WebKit::partsForItem):
(WebKit::addPartsForText):
(WebKit::addPartsForItem):
(WebKit::addTextRepresentationRecursive):

Implement the bulk of the algorithm here. This recursively traverses extraction results, adding
results to the `aggregator` in the process. This takes an async filtering callback (which returns a
`NativePromise`) which applies to all extracted text items. While any of these promises is pending,
the `aggregator` will be kept alive.

(WebKit::convertToText):
* Source/WebKit/Shared/TextExtractionToStringConversion.h: Copied from Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.h.
(WebKit::convertToText):
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _activeNativeMenuItemTitles]):

Make this return a list of item titles, now that we no longer use `WKTextExtractionPopupMenu`.

(-[WKWebView _debugTextWithConfiguration:completionHandler:]):

Reimplement this in terms of `-_requestTextExtractionInternal:completion:`; instead of getting a
`WKTextExtractionResult` and converting that into a text representation, we now go directly from
`TextExtraction::Item` to `NSString *`.

We also reimplement the async filtering logic here (i.e. passing text through
`WebKit::TextExtractionFilter` and `-_validateText:`) using the `TextExtractionFilterCallback` in
`convertToText`, which allows us to delete the now-unused helpers to recursively filter out text in
the ObjC extraction result tree, in `WKTextExtractionUtilities.mm`.

(-[WKWebView _requestTextExtractionInternal:completion:]):

Pull common logic to obtain an `TextExtraction::Item` out into a separate helper method.

(-[WKWebView _requestTextExtraction:completionHandler:]):
(-[WKWebView _validateText:inNode:completionHandler:]):
(-[WKWebView _popupMenuForTextExtractionResults]): Deleted.

Remove `WKTextExtractionPopupMenu` entirely, now that debug text extraction no longer depends on
producing a `WKTextExtractionResult` and asking it for a text representation.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.swift:
(WKTextExtractionEventListenerTypes.description): Deleted.
(eventListenerTypesAsArray(_:)): Deleted.
(String.escaped): Deleted.
(WKTextExtractionItem.textRepresentationRecursive(_:)): Deleted.
(WKTextExtractionItem.textRepresentationParts): Deleted.
(WKTextExtractionContainerItem.textRepresentationParts): Deleted.
(WKTextExtractionContentEditableItem.textRepresentationParts): Deleted.
(WKTextExtractionTextFormControlItem.textRepresentationParts): Deleted.
(WKTextExtractionLinkItem.textRepresentationParts): Deleted.
(WKTextExtractionTextItem.textRepresentationParts): Deleted.
(WKTextExtractionScrollableItem.textRepresentationParts): Deleted.
(WKTextExtractionSelectItem.textRepresentationParts): Deleted.
(WKTextExtractionImageItem.textRepresentationParts): Deleted.
(WKTextExtractionPopupMenu.textRepresentation): Deleted.
(WKTextExtractionResult.textRepresentation): Deleted.

Delete now-unused code (see `TextExtractionToStringConversion.cpp` above).

* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtractionInternal.h:
* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.h:
* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.mm:
(WebKit::filterTextRecursive): Deleted.
(WebKit::filterText): Deleted.

Moved into `WKWebView.mm` (see above).

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/301002@main
